File: src/util/main/sp/it/util/collections/map/PrefListMap.java
package sp.it.util.collections.map;

import kotlin.jvm.functions.Function1;
import sp.it.util.collections.list.PrefList;

public class PrefListMap<E, K> extends CollectionMap<E,K,PrefList<E>> {

	public PrefListMap(Function1<E,K> keyMapper) {
		super(PrefList::new, keyMapper);
	}

	public void accumulate(E e, boolean pref) {
		K k = keyMapper.invoke(e);
		PrefList<E> c = computeIfAbsent(k, key -> cacheFactory.get());
		c.addPreferred(e, pref);
	}

}

File: src/util/main/sp/it/util/collections/map/PrefListMap.kt
package sp.it.util.collections.map

import java.util.function.Supplier
import sp.it.util.collections.list.PrefList
import sp.it.util.functional.invoke

class PrefListMap<E: Any, K>(keyMapper: (E) -> K): CollectionMap<E, K, PrefList<E>>(Supplier { PrefList() }, keyMapper) {

   fun accumulate(e: E, pref: Boolean) {
      val k = keyMapper(e)
      val c = computeIfAbsent(k) { cacheFactory() }
      c.addPreferred(e, pref)
   }

}